Remixer should fail-fast if Application is not implemented

Context: http://codereview.cc/D1976#40068

I tried integrating Remixer from memory, and I didn't subclass Application correct (to call the remixer initialization code). This should fail-fast so the user knows they forgot to do something. For example, I just realized that my previous integration with `rebound` has the same issue. I did [subclass application](https://github.com/material-motion/family-rebound-android/blob/develop/sample/src/main/java/com/google/android/material/motion/family/rebound/sample/ReboundApplication.java) but I did not add it to my [AndroidManifest](https://github.com/material-motion/family-rebound-android/blob/develop/sample/src/main/AndroidManifest.xml#L20). I would have never known this would be a problem.

Here's the current exception that's thrown
```
12-15 21:07:09.245 6276-6276/com.google.android.material.motion.family.rebound.sample E/AndroidRuntime: FATAL EXCEPTION: main
                                                                                                        Process: com.google.android.material.motion.family.rebound.sample, PID: 6276
                                                                                                        java.lang.RuntimeException: Unable to start activity ComponentInfo{com.google.android.material.motion.family.rebound.sample/com.google.android.material.motion.family.rebound.sample.MainActivity}: java.lang.IllegalArgumentException: dataType cannot be unset for RemixerItem
                                                                                                            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2665)
                                                                                                            at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2726)
                                                                                                            at android.app.ActivityThread.-wrap12(ActivityThread.java)
                                                                                                            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1477)
                                                                                                            at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                                                            at android.os.Looper.loop(Looper.java:154)
                                                                                                            at android.app.ActivityThread.main(ActivityThread.java:6119)
                                                                                                            at java.lang.reflect.Method.invoke(Native Method)
                                                                                                            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
                                                                                                            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
                                                                                                         Caused by: java.lang.IllegalArgumentException: dataType cannot be unset for RemixerItem
                                                                                                            at com.google.android.libraries.remixer.RemixerItem$Builder.checkBaseFields(RemixerItem.java:150)
                                                                                                            at com.google.android.libraries.remixer.RangeVariable$Builder.build(RangeVariable.java:206)
                                                                                                            at com.google.android.material.motion.family.rebound.sample.MainActivity_RemixerBinder.bindInstance(MainActivity_RemixerBinder.java:29)
                                                                                                            at com.google.android.material.motion.family.rebound.sample.MainActivity_RemixerBinder.bindInstance(MainActivity_RemixerBinder.java:13)
                                                                                                            at com.google.android.libraries.remixer.annotation.RemixerBinder.bind(RemixerBinder.java:37)
                                                                                                            at com.google.android.material.motion.family.rebound.sample.MainActivity.onCreate(MainActivity.java:59)
                                                                                                            at android.app.Activity.performCreate(Activity.java:6679)
                                                                                                            at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1118)
                                                                                                            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2618)
                                                                                                            at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2726) 
                                                                                                            at android.app.ActivityThread.-wrap12(ActivityThread.java) 
                                                                                                            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1477) 
                                                                                                            at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                                                            at android.os.Looper.loop(Looper.java:154) 
                                                                                                            at android.app.ActivityThread.main(ActivityThread.java:6119) 
                                                                                                            at java.lang.reflect.Method.invoke(Native Method) 
                                                                                                            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886) 
                                                                                                            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776) 
```

It's super confusing. Please throw an exception that reminds the client to implement Application correctly!

--- a/README.md
+++ b/README.md
@@ -68,7 +68,7 @@ Notice the dependency on `remixer_annotation` is a `provided` clause instead of 
 ### Global remixer set up
 If you have not subclassed the application class it is recommended you do it since this is a one-time global initialization.
 
-In your application class you need to call the Remixer initialization method.
+In your application class you need to call the Remixer initialization method. Remember this application class MUST be (declared in your Android Manifest)[https://developer.android.com/guide/topics/manifest/application-element.html#nm].
 
 ```java
 import android.app.Application;

--- a/remixer_core/src/main/java/com/google/android/libraries/remixer/Remixer.java
+++ b/remixer_core/src/main/java/com/google/android/libraries/remixer/Remixer.java
@@ -94,7 +94,7 @@ public class Remixer {
     return registeredDataTypes.get(name);
   }
 
-  public static Collection<DataType> getRegisteredDataType() {
+  public static Collection<DataType> getRegisteredDataTypes() {
     return registeredDataTypes.values();
   }
 
@@ -149,6 +149,14 @@ public class Remixer {
    */
   @SuppressWarnings("unchecked")
   public void addItem(RemixerItem remixerItem) {
+    if (!registeredDataTypes.containsKey(remixerItem.getDataType().getName())) {
+      throw new IllegalStateException(String.format(
+          Locale.getDefault(),
+          "There is no registered data type that matches %s. Are you sure you ran "
+          + "RemixerInitialization.initRemixer in your application class? See the Remixer README "
+          + "for detailed instructions. If this is a custom data type you have to manually add it.",
+          remixerItem.getDataType().getName()));
+    }
     List<RemixerItem> listForKey = getOrCreateItemList(remixerItem.getKey(), keyMap);
     for (RemixerItem existingItem : listForKey) {
       if (remixerItem.getContext() != null

--- a/remixer_core/src/main/java/com/google/android/libraries/remixer/annotation/RemixerBinder.java
+++ b/remixer_core/src/main/java/com/google/android/libraries/remixer/annotation/RemixerBinder.java
@@ -16,6 +16,9 @@
 
 package com.google.android.libraries.remixer.annotation;
 
+import com.google.android.libraries.remixer.Remixer;
+import java.util.Locale;
+
 /**
  * Used to bind activities to their implicitly generated variables and triggers.
  */
@@ -28,6 +31,14 @@ public class RemixerBinder {
    */
   @SuppressWarnings("ClassNewInstance")
   public static <T> void bind(T activity) {
+    if (Remixer.getRegisteredDataTypes().isEmpty()) {
+      throw new IllegalStateException(
+          "There are no registered data types for remixer. This indicates that you have not "
+          + " initialized Remixer at all and it will fail in runtime. Please run "
+          + "RemixerInitialization.initRemixer in your application class. See the Remixer README "
+          + "for detailed instructions. ");
+    }
+
     try {
       Class<?> bindingClass =
           Class.forName(activity.getClass().getCanonicalName() + "_RemixerBinder");

--- a/remixer_core/src/main/java/com/google/android/libraries/remixer/serialization/StoredVariable.java
+++ b/remixer_core/src/main/java/com/google/android/libraries/remixer/serialization/StoredVariable.java
@@ -212,7 +212,7 @@ public class StoredVariable<T> {
    */
   static StoredVariable fromRemixerItem(RemixerItem item) {
     StoredVariable storedVariable = null;
-    for (DataType type : Remixer.getRegisteredDataType()) {
+    for (DataType type : Remixer.getRegisteredDataTypes()) {
       try {
         storedVariable = type.getConverter().fromRemixerItem(item);
         break;

--- a/remixer_core/src/main/java/com/google/android/libraries/remixer/serialization/StoredVariableSerializer.java
+++ b/remixer_core/src/main/java/com/google/android/libraries/remixer/serialization/StoredVariableSerializer.java
@@ -46,7 +46,7 @@ class StoredVariableSerializer
     JsonPrimitive jsonDataType = object.getAsJsonPrimitive(StoredVariable.DATA_TYPE);
     String dataType = jsonDataType.getAsString();
     StoredVariable variable = null;
-    for (DataType type : Remixer.getInstance().getRegisteredDataType()) {
+    for (DataType type : Remixer.getInstance().getRegisteredDataTypes()) {
       if (dataType.equals(type.getName())) {
         variable = type.getConverter().deserialize(json);
         break;
@@ -64,7 +64,7 @@ class StoredVariableSerializer
   public JsonElement serialize(
       StoredVariable src, Type typeOfSrc, JsonSerializationContext context) {
     ValueConverter converter = null;
-    for (DataType type : Remixer.getInstance().getRegisteredDataType()) {
+    for (DataType type : Remixer.getInstance().getRegisteredDataTypes()) {
       if (src.dataType.equals(type.getName())) {
         converter = type.getConverter();
         break;

--- a/remixer_core/src/test/java/com/google/android/libraries/remixer/FailfastTest.java
+++ b/remixer_core/src/test/java/com/google/android/libraries/remixer/FailfastTest.java
@@ -23,6 +23,7 @@ import org.junit.runners.JUnit4;
 
 @RunWith(JUnit4.class)
 public class FailfastTest {
+  
   @Test(expected = IllegalStateException.class)
   public void binderFailsFast() {
     RemixerBinder.bind(this);

--- a/remixer_core/src/test/java/com/google/android/libraries/remixer/FailfastTest.java
+++ b/remixer_core/src/test/java/com/google/android/libraries/remixer/FailfastTest.java
@@ -1,0 +1,35 @@
+/*
+ * Copyright 2016 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.libraries.remixer;
+
+import com.google.android.libraries.remixer.annotation.RemixerBinder;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class FailfastTest {
+  @Test(expected = IllegalStateException.class)
+  public void binderFailsFast() {
+    RemixerBinder.bind(this);
+  }
+
+  @Test(expected = IllegalStateException.class)
+  public void remixerAddFailsFast() {
+    Remixer.getInstance().addItem(new Trigger.Builder().setContext(this).setKey("key").build());
+  }
+}

